### PR TITLE
Moved VMCI code to one place and added distinct error return codes

### PR DESCRIPTION
**Pull request by msterin**
First pass on vmci errors clean up:
* all vmci C code is now in esx_service/vmci , so no duplicate .h files (was leftover from prototyping)
* distinct code is returned instead of -1
* client log and Dockert command line are getting proper message

TBD in next PR:
* proper error messages log on the server
* bubble up errno and extra details (currently fprintf(stderr) - not really visible in service logs)

Tested: 
manual : stopped esx service and checked Docker prints proper message
CI

```
msterin-m01:docker-volume-vsphere msterin$ ssh root@$VM docker volume ls
DRIVER              VOLUME NAME
list vmdk: VolumeDriver.List: Failed to communicate with ESX (failed to connect to vSocket)
msterin-m01:docker-volume-vsphere msterin$ 
```

### CODE REVIEW

File: esx_service/vmci/connection_errors.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      { CONN_FAILED_CONTENT_RECIEVE, "failed to receive content"},
+      { CONN_MALLOC_FAILED, "failed malloc"},
+      { CONN_BUF_TOO_SMALL, "request buffer is too small"},
+      { CONN_LEN_MISMATCH, "message length mismatch"},
+      { CONN_BAD_BE_NAME, "bad back end name"},
+
+      { CONN_SUCCESS, NULL}
+   };
+
+   for (i=0; info[i].msg != NULL;  i++) {
```
Comment (pdhamdhere) [inline esx_service/vmci/connection_errors.h]: Why do you have to loop rather than returning info[i].msg?

Comment (msterin) [inline esx_service/vmci/connection_errors.h]: For reliabilit -  I do not want to rely on enum values being the correct indexes, especially if they shift/soft/etc.. the definition.
I can craft something with macros but the cost of the loop is negligible

File: vmdk_plugin/vmdkops/esx_vmdkcmd.go [hunk 2]
```diff
@@ -84,10 +84,11 @@ func (vmdkCmd EsxVmdkCmd) Run(cmd string, name string, opts map[string]string) (
 	// connect, send command, get reply, disconnect - all in one shot
 	ret := C.Vmci_GetReply(C.int(vmciEsxPort), cmdS, beS, ans)
 
-	if ret != 0 {
-		msg := "Failed to connect to ESX over vsocket"
+	if ret != nil {
+		msg := fmt.Sprintf("Failed to communicate with ESX (%v)",
+			C.GoString(ret))
 		// TODO: per-errcode msg, when vmci_client.c err return is fixed. Issue #206
```
Comment (pdhamdhere) [inline vmdk_plugin/vmdkops/esx_vmdkcmd.go]: Remove TODO in this change?

Comment (msterin) [inline vmdk_plugin/vmdkops/esx_vmdkcmd.go]: Sure - when errno is bubbled up (next PR hopefully)

**Comment by pdhamdhere** [comment]
With this PR, we know failing API but we still won't know errno returned by VMCI? Isn't that most crucial information for debugging vsocket issues?

**Comment by pdhamdhere** [comment]
Ignore my last comment. I see that you have already mentioned in description that it will be handled in next PR. LGTM.

File: esx_service/vmci/connection_errors.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+// Errors on vmci command communication channel
+
+#ifndef _CONNECTION_ERRORS_H_
+#define _CONNECTION_ERRORS_H_
+
+typedef enum VMCI_ConnectionError {
+   CONN_SUCCESS = 0,
+
+   // Misc. protocol errors. When changing, also update info[] below
+   CONN_FAILED_VMCI_ADDRESS_FAMILY_GET,
```
Comment (govint) [inline esx_service/vmci/connection_errors.h]: The error should reflect the error vs. the function that failed. The original was correct - CONN_NO_VMCI_ADDRESS_FAMILY., New one can be CONN_FAILED_NO_VMCI_ADDRESS_FAMILY.

File: esx_service/vmci/connection_errors.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      { CONN_FAILED_VSOCKET_ACCEPT, "failed on accept on vSocket"},
+      { CONN_FAILED_TO_CONNECT, "failed to connect to vSocket"},
+      { CONN_FAILED_SOCKADDR_GET, "failed sockadr get"},
+      { CONN_FAILED_MAGIC_SEND, "failed MAGIC send"},
+      { CONN_FAILED_LEN_SEND, "failed LEN send"},
+      { CONN_FAILED_CONTENT_SEND, "failed content send"},
+      { CONN_FAILED_MAGIC_RECIEVE, "failed MAGIC receive"},
+      { CONN_FAILED_LEN_RECEIVE, "failed LEN receive"},
+      { CONN_FAILED_CONTENT_RECIEVE, "failed to receive content"},
+      { CONN_MALLOC_FAILED, "failed malloc"},
```
Comment (govint) [inline esx_service/vmci/connection_errors.h]: Can be CONN_FAILED_MALLOC - to match the CONN_FAILED prefix like the others.

File: esx_service/vmci/connection_errors.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      { CONN_BUF_TOO_SMALL, "request buffer is too small"},
+      { CONN_LEN_MISMATCH, "message length mismatch"},
+      { CONN_BAD_BE_NAME, "bad back end name"},
+
```
Comment (govint) [inline esx_service/vmci/connection_errors.h]: For the above three error should be CONN_FAILED_......., to match the error list.

File: esx_service/vmci/connection_errors.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+   // Bad communication back end name (internal usage only)
+   CONN_BAD_BE_NAME
+} VMCI_ConnectionError;
+
+// gets error string for a connection error
+static const char*
+VMCI_ErrorStr(VMCI_ConnectionError id)
+{
+   int i;
+   struct {
```
Comment (govint) [inline esx_service/vmci/connection_errors.h]: static?

Comment (govint) [inline esx_service/vmci/connection_errors.h]: nit - this structure could be outside this function

File: vmdk_plugin/vmdkops/esx_vmdkcmd.go [hunk 7]
```diff
@@ -84,10 +84,11 @@ func (vmdkCmd EsxVmdkCmd) Run(cmd string, name string, opts map[string]string) (
...
-		log.Warnf("%s (return %d)", msg, ret)
+		log.Warnf(msg)
```
Comment (govint) [inline vmdk_plugin/vmdkops/esx_vmdkcmd.go]: For debug the ret value is useful to have.

File: esx_service/vmci/vmci_client.c [hunk 8]
```diff
@@ -200,13 +200,13 @@ vsock_init(be_sock_id *id, int cid, int port)
    id->addr.svm_cid = cid;
    id->addr.svm_port = port;
    ret = connect(sock, (const struct sockaddr *) &id->addr, sizeof id->addr);
-   if (ret == -1) {
+   if (ret != 0) {
```
Comment (andrewjstone) [inline esx_service/vmci/vmci_client.c]: `ret != CONN_SUCCESS` ?

**Comment by andrewjstone** [comment]
I agree with @pdhamdhere and @govint on enum naming. Other than that LGTM. Nice work @msterin

File: esx_service/vmci/vmci_client.c (hunk 8 above)
Comment (msterin) [inline esx_service/vmci/vmci_client.c]: No, this is 0 from connect(), not our internal connection flags

File: vmdk_plugin/vmdkops/esx_vmdkcmd.go (hunk 7 above)
Comment (msterin) [inline vmdk_plugin/vmdkops/esx_vmdkcmd.go]: Yes, and it there in the msg

File: esx_service/vmci/connection_errors.h (hunk 6 above)
Comment (msterin) [inline esx_service/vmci/connection_errors.h]: Why? It's perfectly fine inside. It should be static though to save on unit (cheap but still...)  will fix

File: esx_service/vmci/connection_errors.h (hunk 4 above)
Comment (msterin) [inline esx_service/vmci/connection_errors.h]: Ok

File: esx_service/vmci/connection_errors.h (hunk 5 above)
Comment (msterin) [inline esx_service/vmci/connection_errors.h]: Nope, these are diliberately different

**Comment by msterin** [comment]
Thanks for the review , folks .
I'll address nits and naming suggestions
In the follow up PR.

**Pull request merged**